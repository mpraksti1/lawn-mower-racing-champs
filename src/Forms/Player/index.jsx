import React from 'react';
import PropTypes from 'prop-types';
import * as Yup from 'yup';
import { withAlert } from 'react-alert';
import { Formik, Field, Form } from 'formik';
import PersonIcon from 'react-icons/lib/md/person';
import StarIcon from 'react-icons/lib/md/star';
import HandIcon from 'react-icons/lib/md/pan-tool';
import BaseForm from '../BaseForm';
import Button from '../../Elements/Button';
import Text from '../../Elements/Text';

const PlayerrSchema = Yup.object().shape({
  first_name: Yup.string()
    .min(2, 'Name must be at least 2 characters long')
    .required('Required'),
  last_name: Yup.string()
    .min(2, 'Name must be at least 2 characters long')
    .required('Required'),
  rating: Yup.number()
    .min(1)
    .max(10000)
    .required('Required'),
  handedness: Yup.string()
    .oneOf(['right', 'left'], 'Invalid Handedness')
    .required('Required'),
});

const MyForm = ({ touched, errors, isSubmitting }) => (
  <div>
    <Text xlg thin sans spaceAround block>Add a racer:</Text>
    <Form>
      <div className="field-wrapper">
        <label htmlFor="firstName">
          <Text xsm thin sans spaceBelow block>First Name</Text>
        </label>
        <div className="icon-input">
          <PersonIcon />
          <Field type="text" name="first_name" placeholder="Alex" id="firstName" />
        </div>
        {errors.first_name && touched.first_name && (
          <Text block spaceAbove sm red className="field-error">{errors.first_name}</Text>
        )}
      </div>
      <div className="field-wrapper">
        <label htmlFor="lastName">
          <Text xsm thin sans spaceBelow block>Last Name</Text>
        </label>
        <div className="icon-input">
          <PersonIcon />
          <Field type="text" name="last_name" placeholder="Johnson" id="lastName" />
        </div>
        {errors.last_name && touched.last_name && (
          <Text block spaceAbove sm red className="field-error">{errors.last_name}</Text>
        )}
      </div>
      <div className="field-wrapper">
        <label htmlFor="rating">
          <Text xsm thin sans spaceBelow block>Rating</Text>
        </label>
        <div className="icon-input">
          <StarIcon />
          <Field type="text" name="rating" id="rating" placeholder="8000" />
        </div>
        {errors.rating && touched.rating && (
          <Text block spaceAbove sm red className="field-error">{errors.rating}</Text>
        )}
      </div>
      <div className="field-wrapper">
        <label htmlFor="handedness">
          <Text xsm thin sans spaceBelow block>Handedness</Text>
        </label>
        <div className="icon-input">
          <HandIcon />
          <Field component="select" name="handedness" id="handedness">
            <option value="right">Right</option>
            <option value="left">Left</option>
          </Field>
        </div>
        {errors.handedness && touched.handedness && (
          <Text block spaceAbove sm red className="field-error">{errors.handedness}</Text>
        )}
      </div>
      <Button outline type="submit" disabled={isSubmitting} id="create">Create</Button>
    </Form>
  </div>
);

export const PlayerForm = props => (
  <BaseForm>
    <Formik
      initialValues={{
        first_name: '',
        last_name: '',
        rating: '',
        handedness: 'right',
      }}
      validationSchema={PlayerrSchema}
      onSubmit={async (player) => {
        const { token } = window.localStorage;
        try {
          const { createPlayer } = props;
          const playerResult = await createPlayer(player, token);
          if (playerResult.success) {
            props.history.push('/roster');
          } else {
            props.alert.error(playerResult.error.message);
          }
        } catch (error) {
          props.alert.error(error.message);
        }
      }}
      component={MyForm}
    />
  </BaseForm>
);

MyForm.propTypes = {
  // eslint-disable-next-line
  errors: PropTypes.object.isRequired,
  // eslint-disable-next-line
  touched: PropTypes.object.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
};

PlayerForm.propTypes = {
  // eslint-disable-next-line
  history: PropTypes.object.isRequired,
  // eslint-disable-next-line
  alert: PropTypes.object.isRequired,
  createPlayer: PropTypes.func.isRequired,
};

export default withAlert(PlayerForm);
