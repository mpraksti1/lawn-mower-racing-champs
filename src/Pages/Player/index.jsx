import React from 'react';
import PropTypes from 'prop-types';
import BGVideo from '../../Components/BGVideo';
import FormModal from '../../Components/FormModal';

const Player = (props) => {
  const { isRequesting, createPlayer } = props;

  return (
    <div className="player-page">
      <BGVideo dark url="/assets/video/bgvid.mp4" />
      <FormModal createPlayer={createPlayer} isRequesting={isRequesting} />
    </div>
  );
};

Player.propTypes = {
  isRequesting: PropTypes.bool.isRequired,
  createPlayer: PropTypes.func.isRequired,
};

export default Player;
