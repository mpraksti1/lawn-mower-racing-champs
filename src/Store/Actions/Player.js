import {
  CREATE_PLAYER,
  DELETE_PLAYER,
  RETRIEVE_PLAYERS,
  ASYNC_REQUESTED,
  ASYNC_COMPLETE,
} from './Types';
import to from '../../awaitToJs';

const api = 'https://players-api.developer.alchemy.codes/api';


// Requests

async function retrievePlayersRequest(token) {
  const response = await fetch(`${api}/players`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return response.json();
}

async function createPlayerRequest(player, token) {
  const data = JSON.stringify(player);

  const response = await fetch(`${api}/players`, {
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    method: 'POST',
    body: data,
  });
  return response.json();
}

async function deletePlayerRequest(id, token) {
  const response = await fetch(`${api}/players/${id}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
    method: 'DELETE',
  });
  return response.json();
}


// Thunks

export const retrievePlayers = player => async (dispatch) => {
  dispatch({ type: ASYNC_REQUESTED });

  let err;
  let callResponse;

  const onError = (error) => {
    throw new Error(error.message);
  };

  const onSuccess = (response) => {
    dispatch({
      type: RETRIEVE_PLAYERS,
      payload: response.players,
    });
    return response;
  };

  // eslint-disable-next-line prefer-const
  [err, callResponse] = await to(retrievePlayersRequest(player));

  if (callResponse) {
    dispatch({ type: ASYNC_COMPLETE });
    return onSuccess(callResponse);
  }

  if (err) {
    dispatch({ type: ASYNC_COMPLETE });
    return onError(err);
  }

  return onError({ message: 'Whoops, there was a problem with your request' });
};

export const createPlayer = (player, token) => async (dispatch) => {
  dispatch({ type: ASYNC_REQUESTED });

  let err;
  let callResponse;

  const onError = (error) => {
    throw new Error(error.message);
  };

  const onSuccess = (response) => {
    dispatch({
      type: CREATE_PLAYER,
      payload: response,
    });
    return response;
  };
  // eslint-disable-next-line prefer-const
  [err, callResponse] = await to(createPlayerRequest(player, token));

  if (callResponse) {
    dispatch({ type: ASYNC_COMPLETE });
    return onSuccess(callResponse);
  }

  if (err) {
    dispatch({ type: ASYNC_COMPLETE });
    return onError(err);
  }

  return onError({ message: 'Whoops, there was a problem with your request' });
};

export const deletePlayer = (id, token) => async (dispatch) => {
  dispatch({ type: ASYNC_REQUESTED });

  let err;
  let callResponse;

  const onError = (error) => {
    throw new Error(error.message);
  };
  const onSuccess = (data) => {
    dispatch({
      type: DELETE_PLAYER,
    });
    return data;
  };

  // eslint-disable-next-line prefer-const
  [err, callResponse] = await to(deletePlayerRequest(id, token));

  if (callResponse) {
    dispatch({ type: ASYNC_COMPLETE });
    return onSuccess(callResponse);
  }

  if (err) {
    dispatch({ type: ASYNC_COMPLETE });
    return onError(err);
  }

  return onError({ message: 'Whoops, there was a problem with your request' });
};
