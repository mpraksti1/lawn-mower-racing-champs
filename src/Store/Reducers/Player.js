import {
  CREATE_PLAYER,
  RETRIEVE_PLAYERS,
} from '../Actions/Types';

const initialState = {
  players: [],
};

export default (state = initialState, action) => {
  switch (action.type) {
    case CREATE_PLAYER:
      return {
        ...state,
        players: [...state.players, action.payload ]
      }

    case RETRIEVE_PLAYERS:
      return {
        ...state,
        players: [ ...action.payload ]
      }

    default:
      return state
  }
}

