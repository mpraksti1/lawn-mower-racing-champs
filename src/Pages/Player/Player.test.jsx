import Player from '.';

describe('Player page component', () => {
  it('renders without crashing', () => {
    shallow(<Player createPlayer={() => {}} isRequesting={false} />);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<Player createPlayer={() => {}} isRequesting={false} />);

    expect(wrapper).toMatchSnapshot();
  });
});
