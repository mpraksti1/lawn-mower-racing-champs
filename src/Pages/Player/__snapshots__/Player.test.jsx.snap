// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player page component matches snapshot 1`] = `
<div
  className="player-page"
>
  <BGVideo
    dark={true}
    url="/assets/video/bgvid.mp4"
  />
  <withRouter(FormModal)
    createPlayer={[Function]}
    isRequesting={false}
  />
</div>
`;
